### ISSUE TITLE
If --package is specified, "@version" should be parsed on that (and only that)

### ISSUE THREAD

**Issue by stuartpb** [zkat/npx#1]
Long story short, if I want to execute a script in a package that has an `@` in the filename (like maybe its bin scripts follow a similar naming convention to systemd unit files or something), I should be able to do `npx -p like-units script@port.service` and not worry that `port.service` might be interpreted as a tag name for the `like-units` package (and, of course, if I want to specify a specific version of the package, I'd do it as `npx -p the-package@6.2.8 the-script` and not `npx -p the-package the-script@6.2.8`).

This also draws a guaranteed wall between the command to be run and the arguments to `npx` so long as the package name is explicitly defined (one that can be enforced with `--` a la other command-running programs' arguments, such as `ssh`); ideally, I should be able to drop `npx --package $mypackage@$whatever -- "$@"` as an alias into one of my scripts knowing I don't need to worry about some esoteric `mypackage`'s `-1` script being misinterpreted due to the `npx` layer.

**Comment by zkat** [comment]
Thanks for the suggestion! This was almost exactly what I was doing already buuuut it works now:

<img width="416" alt="screen shot 2017-05-29 at 17 57 41" src="https://cloud.githubusercontent.com/assets/17535/26564646/7321d7f0-4498-11e7-97fd-a198a9173cbc.png">

This'll be included when I actually release npx. To be clear: I split out everything after `<command>` and it is interpreted entirely by `<command>` at that point. Only things _left_ of the command are parsed by npx.

**Comment by stuartpb** [comment]
:+1:

Is `--` to explicitly signify that the next argument is `<command>` implemented?

**Comment by zkat** [comment]
@stuartpb oh. No, not that. I don't think I would do that. The first non-option argument is always interpreted as a command, and it will be used exactly as-is if you provide a `-p` option. `npx` itself accepts no variadic argument, so the parser is able to disambiguate this easily enough.

Therefore, `npx -p $mypackage@$myversion $@` will be interpreted the way you're describing, unless I misunderstood something.

**Comment by stuartpb** [comment]
> The first non-option argument is always interpreted as a command, and it will be used exactly as-is if you provide a `-p` option. `npx` itself accepts no variadic argument, so the parser is able to disambiguate this easily enough.

Well, no, because you can have a command *with the same name as* an option. It's perfectly valid to have a bin script called `-c` - it might not be *advisable* (due to the potential confusion in scenarios just like this), but it can make sense in certain contexts (for instance, for a package that, itself, uses scripts *for arguments to the script itself* - imagine a project that ran its `crispify` script when called as `foobar crispify`, then wanted to add `foobar -c` as a variant alias for drop-in intercompatibility with a similar tool). Without `--` to signal "the next argument is the beginning of the command", there's no way to distinguish this unambiguously.

**Comment by stuartpb** [comment]
Illustrated:

```bash
npx -p foo -c 'require("alert")(1)'
# executes 'require("alert")(1)' in an npm run-script environment

npx -p foo -- -c 'require("alert")(1)'
# runs the script '-c' with 'require("alert")(1)' as its first argument
```

**Comment by zkat** [comment]
@stuartpb Everything in this issue has (I believe) been implemented as of `npx@1.1.0`. Thanks for all the ideas! Please do keep them coming. I'll close this issue in particular, though. (and if what you described here doesn't work, it might be a bug, and worth fixing).

**Comment by stuartpb** [comment]
:100:
